[Chia Wen Ling] iP

cwenling opened the pull request.

Commit:
Add Gradle support

cwenling committed:
completed level 1

cwenling committed:
Completed Level-2: Add, List

cwenling committed:
completed Level-3: Mark as Done

cwenling committed:
completed Level-4: ToDos, Events, Deadlines

cwenling committed:
Added Automated Text UI Testing

cwenling committed:
Completed Level-5: Handle Errors

Changed design of duke greeting.
Updated text-ui-test to test errors.

cwenling committed:
Completed Level-6:Delete

Updated text-ui-test.

cwenling committed:
Task class: change isDone symbols

Tick and cross were appearing in runtest as question marks.
Changed symbol used to represent isDone and !isDone to x and o.

cwenling committed:
Modify runtest path and added UTF-8 support

Runtest.bat file was using absolute path.
This will not work on other computers.
Let's change runtest.bat file to use relative path.

Revert to using tick and cross to represent isDone.
Added UTF-8 Encoding to correctly represent symbols in runtest.

cwenling committed:
Add Level-7: Save

Tasks were stored in temporary variables.
This requires the user to restart from a blank slate whenever they use
Duke.
Let's save the tasks to the hard disk whenever the user ends the
program.

cwenling committed:
Add Level-8: Dates and Times

Date strings were not processed.
Parsing the dates will allow for more functionality such as sorting by
date in the future. Also enforces actual dates to be input.
Let's check for the correctness of the input date before storing the
date.

cwenling committed:
Merge branch 'branch-Level-7'

cwenling committed:
Merge branch 'branch-Level-8'

cwenling committed:
Add Ui class

Duke was starting to get cluttered.
Add a user interface class to deal with interactions with the user.
Modify Duke to use Ui methods.

cwenling committed:
Added Storage class

Storage class used to manage reading from, and writing to data file.

cwenling committed:
Add Tasklist class

Abstracted some functionality of creating the tasklist.

cwenling committed:
Add Parser class and Commands classes

Parser class deals with making sense of user commands.
Commands classes include AddCommand, DoneCommand, DeleteCommand,
ListCommand, ExitCommand, and UnknownCommand.

cwenling committed:
Change LocalDate to LocalDateTime

Deadline and Event timings were represented using dates without timing.
This is unintuitive.
Let's allow the user to add the timing within the day.

cwenling committed:
Add A-MoreOOP: Use More OOP

OOP allows for greater extensibility.
Touch up the Parser, Tasklist, and Commands classes to be more OOP.

cwenling committed:
Add A-Packages: Organize into Packages

Having all files in one directory makes the working directory cluttered.
Let's package it up to control access to different classes.

cwenling committed:
Remove unused method and expand imports

cwenling committed:
Add ability to save tasks upon command

Tasks were saved only upon giving an ExitCommand.
This meant that tasks would not be saved if user quits the program
abruptly. We should not expect the user to quit the program properly
everytime.
Let's allow the program to save the tasklist to the hard disk whenever a
task is changed, added, or deleted.

cwenling committed:
Re-package Duke and refactor code

cwenling committed:
Add A-JUnit: Add JUnit Tests

Add JUnit tests to test the behavior of the code.
Add 2 test methods to test ExitCommand and Deadline classes.

Change Deadline and Event time formatting.
Removed exception propagation in Tasklist class, to ignore exception
instead. Exception is unlikely to occur if code behavior is consistent
anyway.

cwenling committed:
Add A-Jar: Create a JAR File

Package the app as an executable JAR file so that it can be distributed
easily.

cwenling committed:
Add unit test for Task class

cwenling committed:
Add javadocs for *Command classes

cwenling committed:
Add javadocs for Ui, Parser, Storage, DukeException

cwenling committed:
Add javadocs for classes in 'task' package

cwenling committed:
Add A-JavaDoc: JavaDoc

Add JavaDoc comments to Duke class, and edits to JavaDoc comments
in other classes.

JavaDocs helps to improve understandability of code for external
parties.

cwenling committed:
Add A-CodingStandard: Follow the Coding Standard

Tweak the code to comply with the SE-EDU coding standard.

cwenling committed:
Add Level-9: Find

Give users a way to find a task by searching for a keyword.

cwenling committed:
Merge branch 'branch-A-JavaDoc'

cwenling committed:
Merge branch 'branch-A-CodingStandard'

cwenling committed:
Merge branch 'branch-Level-9'

cwenling committed:
Add JavaDoc for FindCommand class

cwenling committed:
Add HelpCommand class

cwenling committed:
Modify text-ui-test to test new features

cwenling committed:
Merge remote-tracking branch 'origin/add-gradle-support'

cwenling committed:
Add A-Gradle: Use Gradle

Automate project builds using Gradle.
Able to build and run Duke using gradle, as well as run unit tests.

cwenling committed:
Add A-CheckStyle: Use CheckStyle

Use checkStyle to detect coding style violations.

cwenling committed:
Add GUI Files

Able to display scroll view and echo.
Unable to refactor code after adding fxml files.

cwenling committed:
Merge branch 'branch-A-Gradle'

cwenling committed:
Merge branch 'branch-A-CheckStyle'

raythx98 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,40 @@
+        tasks.addTasks(task);
+        String echoizer = "\t Got it. I've added this task:\n"
+                + "\t %s\n"
+                + "\t Now you have %d tasks in the list.\n";
```
Is it necessary to wrap it in 3 lines? line 36 should be indented a further 8 spaces, I notice this in other places too

raythx98 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,37 @@
+    /**
+     * Default constructor for Command class.
+     */
+    public Command() {
```
Is this needed?

raythx98 commented on src/main/java/duke/command/UnknownCommand.java:
```diff
@@ -0,0 +1,38 @@
+ */
+public class UnknownCommand extends Command {
+
+    private String gibberish;
```
Is there a better variable name such as gibberishInput ?

raythx98 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,59 @@
+     */
+    public Task(String description) {
+        this.description = description;
+        this.isDone = false;
```
Is the overqualified 'this' necessary? https://se-education.org/guides/conventions/java/index.html under variables

raythx98 commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,83 @@
+
+    private static final String DIVIDER =
+            "\t____________________________________________________________\n";
+    private static final String LOGO =
```
LOL

raythx98 reviewed (commented):
Generally, the code is very neat and well written, I found it difficult to spot any mistakes :)

cwenling committed:
Add Level-10: GUI

Use the JavaFX technology to implement the GUI.
Duke is able to respond vis GUI.

Will refactor to use fxml in next update.

ijavierja commented on src/main/java/duke/command/DoneCommand.java:
```diff
@@ -0,0 +1,44 @@
+ */
+public class DoneCommand extends Command {
+
+    private int i;
```
what does i represent? Can its naming be more specific?

ijavierja commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,54 @@
+     */
+    @Override
+    public String toString() {
+        String by = time.format(DateTimeFormatter.ofPattern("MMM d yyyy HH:mm"));
```
Is the name "by" descriptive enough? Would this confuse the reader?

ijavierja commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,64 @@
+     */
+    @Override
+    public String toString() {
+        String at = timeStart.format(DateTimeFormatter.ofPattern("MMM d yyyy HH:mm"))
```
Is the name "at" descriptive enough?

ijavierja commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,149 @@
+     *
+     * @return Number of items in TaskList.
+     */
+    public int numTasks() {
```
Would recommend using back ArrayList method names since both class have similar functionalities and other readers might be more familiar to ArrayList method names. This would increase readability

cwenling committed:
Refactor GUI using fxml

Added Main and MainWindow classes.
Need to add scrolling functionality.

cwenling committed:
Merge branch 'branch-Level-10'

cwenling committed:
Add styling to GUI

Change font and background colour of chatbot.
Change text styling to fit into window of GUI.

howtoosee commented on src/main/java/duke/command/UnknownCommand.java:
```diff
@@ -2,15 +2,15 @@
 /**
  * Represents an UnknownCommand where user input is unclear.
  */
 public class UnknownCommand extends Command {
```
I think it might be better to use a stronger word as as "InvalidCommand", rather than "unknown", which can give the feeling of the command being unhandled?

howtoosee commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -51,4 +51,4 @@ public String toData() {
         String by = time.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm"));
         return "D" + separator + isDone + separator + super.description + separator + by + "\n";
```
consider using string.format for this

howtoosee commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -120,4 +122,28 @@ public String findTasks(String keyword) {
+     * @return Task which was deleted.
+     */
+    public Task deleteTask(int idx) {
+        Task t = tasks.get(idx - 1);
```
variable name 't' is not very clear

howtoosee commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,100 @@
+     * @param folderPath Path name of duke.Duke text file to be saved.
+     */
+    public Duke(String filePath, String folderPath) {
+        ui = new Ui();
```
i think its better to reference your instance variables starting with the "this" keyword, such as `this.ui` to be clear and prevent variable shadowing

howtoosee commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,100 @@
+        while (!isExit) {
+            try {
+                String fullCommand = ui.readCommand();
+                Command c = Parser.parse(fullCommand);
```
'c' is not a very intuitive variable name. if i rmb correctly, single character names should only be used for scratch variables

howtoosee commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,100 @@
+     * Main method for starting the chat-bot.
+     * @param args Command line arguments, not used.
+     */
+    public static void main(String[] args) {
```
main method as an entry point should be separate from the other class level and instance level methods. consider moving it to the bottom

howtoosee commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,100 @@
+        }
+        boolean isExit = false;
+        try {
+            Command c = Parser.parse(input);
```
'c' as a variable name is not the most intuitive

howtoosee commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,100 @@
+        } catch (DukeException e) {
+            return ui.showError(e.getMessage());
+        }
+        //return " heard: " + input;
```
remove commented out code?

howtoosee commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,40 @@
+    @Override
+    public String execute(TaskList tasks, Ui ui, Storage storage) {
+        tasks.addTasks(task);
+        String echoizer = "  Got it. I've added this task:\n"
```
consider writing this in line with `String.format` for clarity

howtoosee commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,45 @@
+ */
+public class DeleteCommand extends Command {
+
+    private int i;
```
un-intuitive variable name

howtoosee commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,45 @@
+     *
+     * @param i Index of task to be deleted.
+     */
+    public DeleteCommand(int i) {
```
convention to call super() as first line in constructor

howtoosee commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,100 @@
+    public static Command parse(String fullCommand) throws DukeException {
+        Command command;
+
+        if (fullCommand.equals("bye")) {
```
consider using switch rather than if-else. this is slightly hard to read

howtoosee commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,54 @@
+ */
+public class Deadline extends Task {
+
+    protected LocalDateTime time;
```
whats the use of protected here?

howtoosee commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,54 @@
+        try {
+            timeBy = LocalDateTime.parse(by, DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm"));
+        } catch (DateTimeParseException ignored) {
+            return;
```
shouldn't have return statements in constructors

howtoosee commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,64 @@
+    @Override
+    public String toData() {
+        String isDone = super.isDone ? "1" : "0";
+        String separator = "~";
```
since this is repeated for all the events, consider declaring as static final.

howtoosee commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public boolean hasKeyword(String keyword) {
+        return description.contains(keyword) ? true : false;
```
can simply return `description.contains(keyword)` without the need for conditional expression.

howtoosee commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,149 @@
+     * @param taskString Line of text from text file.
+     * @return Task which is specified by the line of text
+     */
+    public Task lineToTask(String taskString) {
```
consider moving this to your parser!

howtoosee commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,149 @@
+     *
+     * @param task Task to be added.
+     */
+    public void addTasks(Task task) {
```
consider renaming to `add` to follow the naming conventions in Collections

howtoosee commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,149 @@
+     * @param idx Position of Task in the list of tasks.
+     * @return Task which was marked as done.
+     */
+    public Task markTask(int idx) {
```
`idx` is not very intuitive :/

howtoosee commented on src/test/java/duke/command/ExitCommandTest.java:
```diff
@@ -0,0 +1,10 @@
+
+public class ExitCommandTest {
+    @Test
+    public void testIsExit() {
```
missing javadoc for test methods

howtoosee reviewed (commented):
Code is clean and readable, with some coding and conventions that are not followed properly. However, goodjoba overall!

cwenling committed:
Fix error in parser

cwenling committed:
Refactor code to comply with naming standards

cwenling committed:
Add missing tags to JavaDocs

cwenling committed:
Add ability to greet user upon starting

cwenling committed:
Remove deprecated methods in Duke

cwenling committed:
Change name to Nite

Refactor class names, package names, and variable names from "Duke" to
"Nite".

cwenling committed:
Change image of cat

cwenling committed:
Add sound effects to chat-bot

Nite will produce a "meow" sound when a command is entered.

cwenling committed:
Add styling to GUI

Remove deprecated methods in Ui class.
Change version of javafx from 11 to 11.0.1.

cwenling committed:
Fixed bug in parser

Fix uncaught exceptions when input invalid DoneCommand or DeleteCommand.
Also, changed profile pictures and removed unused resources.
Changed resource paths to work in Jar files.

cwenling committed:
Add A-Assertions: Use Assertions

Use assert feature to document important assumptions that should
hold at various points in the code.

cwenling committed:
Merge pull request #1 from cwenling/branch-A-Assertions

Add A-Assertions: Use Assertions

cwenling committed:
Add A-CodeQuality: Improve Code Quality

Examine code and refactor to improve code quality where necessary.

cwenling committed:
Add A-Streams: Use Streams

Use the Streams feature of Java to simplify code.

cwenling committed:
Merge pull request #2 from cwenling/branch-A-CodeQuality

Add A-CodeQuality: Improve Code Quality

cwenling committed:
Merge branch 'master' into branch-A-Streams

cwenling committed:
Merge pull request #3 from cwenling/branch-A-Streams

Add A-Streams: Use Streams

cwenling committed:
Add A-CI: Set up CI

Using Continuous Integration (CI).

cwenling committed:
Change location of gradle.yml

cwenling committed:
Remove unused code and imports

cwenling committed:
Fix bug to allow numbering of tasks using streams

cwenling committed:
Add C-Sort: Sorting items managed by the App

Add ability to sort tasks according to type of task, then chronologically,
then alphabetically.

cwenling committed:
Merge pull request #4 from cwenling/branch-C-Sort

Add C-Sort: Sorting items managed by the App

cwenling committed:
Fix formatting bug in JUnit DeadlineTest class

Bug was causing CI automatic checks to fail.
Hope it works now.

Add tags for A-Lambdas and A-Varargs which were added in recent commits.

cwenling committed:
Fix bugs found in input commands

cwenling committed:
Set theme jekyll-theme-midnight

cwenling committed:
Update User Guide for website

cwenling committed:
Merge branch 'master' of https://github.com/cwenling/ip

cwenling committed:
Increase width of Nite GUI

cwenling committed:
Add display of farewell message

cwenling committed:
Add A-UserGuide: Add a User Guide with images

User guide is a guide to all the commands in Nite.

cwenling committed:
Add Ui.png for iP showcase

Ui.png shows the GUI populated with data.
Change to use System.exit instead of Platform.exit because Platform.exit
does not terminate gradle process.

cwenling committed:
Remove unused import

cwenling committed:
Update link of latest release in User Guide

cwenling committed:
Fix bug in parser

cwenling committed:
Modify format of display DateTime

cwenling committed:
Modify Welcome message to be more user-friendly

cwenling committed:
Update README.md
